homeassistant/components/media_player/sonos.py

```diff
@@ -0,0 +1,180 @@
+"""
+homeassistant.components.media_player.sonos
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Provides an interface to Sonos players (via SoCo)
+
+Configuration:
+
+To use SoCo, add something like this to your configuration:
+
+media_player:
+  platform: sonos
+"""
+
+import logging
+import datetime
+
+REQUIREMENTS = ['SoCo==0.11.1']
+
+from homeassistant.components.media_player import (
+    MediaPlayerDevice, SUPPORT_PAUSE, SUPPORT_SEEK, SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_MUTE, SUPPORT_PREVIOUS_TRACK, SUPPORT_NEXT_TRACK,
+    MEDIA_TYPE_MUSIC)
+
+from homeassistant.helpers.event import track_utc_time_change
+from homeassistant.const import (
+    STATE_IDLE, STATE_PLAYING, STATE_PAUSED, STATE_UNKNOWN)
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_SONOS = SUPPORT_PAUSE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE |\
+    SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | SUPPORT_SEEK
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Sets up the Sonos platform. """
+
+    import soco
+    add_devices(SonosDevice(hass, p) for p in soco.discover())
+
+    return True
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-public-methods
+# pylint: disable=abstract-method
+class SonosDevice(MediaPlayerDevice):
+    """ Represents a Sonos device. """
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, hass, player):
+        super(SonosDevice, self).__init__()
+        self._player = player
+        self.update()
+
+        track_utc_time_change(
```

Comment:
The fact that this works is a matter of luck ;-)

Tracking a time changed event will call `callback(now)` with `now` being a datetime object. `update_ha_state` takes 1 argument being `force_update`. If true-ish, it will call `update()`.

The API probably won't change but I prefer if you would do this:
```python
def update_sonos(now):
    self.update_ha_state(True)

track_utc_time_change(…)
```

I prefer a method like this over a lambda so that when the job queue gets big, it has a logical name.